File: sdk-py/smolvm/fleet.py
"""smolvm Python SDK — Fleet.

Multi-sandbox orchestration. Creates N sandboxes, executes commands
across all of them in parallel, and cleans up.
"""

from __future__ import annotations

import asyncio
from typing import Any

from .client import SmolvmHttpClient
from .sandbox import Sandbox
from .types import CreateSandboxOptions, ExecOptions, ExecResult


class SandboxFleet:
    """A fleet of sandboxes for parallel execution."""

    def __init__(self, sandboxes: list[Sandbox], http: SmolvmHttpClient):
        self.sandboxes = sandboxes
        self._http = http

    @property
    def size(self) -> int:
        return len(self.sandboxes)

    @property
    def names(self) -> list[str]:
        return [s.name for s in self.sandboxes]

    async def exec_all(
        self, cmd: str, opts: ExecOptions | None = None
    ) -> list[ExecResult]:
        """Execute a shell command across all sandboxes in parallel."""
        return await asyncio.gather(
            *(s.sh(cmd, opts) for s in self.sandboxes)
        )

    async def exec_each(
        self, cmds: list[str], opts: ExecOptions | None = None
    ) -> list[ExecResult]:
        """Execute different commands on each sandbox in parallel."""
        if len(cmds) != len(self.sandboxes):
            raise ValueError(
                f"Expected {len(self.sandboxes)} commands, got {len(cmds)}"
            )
        return await asyncio.gather(
            *(s.sh(cmd, opts) for s, cmd in zip(self.sandboxes, cmds))
        )

    async def exec_all_raw(
        self, command: list[str], opts: ExecOptions | None = None
    ) -> list[ExecResult]:
        """Execute a raw command array across all sandboxes in parallel."""
        return await asyncio.gather(
            *(s.exec(command, opts) for s in self.sandboxes)
        )

    def at(self, index: int) -> Sandbox:
        """Get a specific sandbox by index."""
        if index < 0 or index >= len(self.sandboxes):
            raise IndexError(
                f"Fleet index {index} out of range (size: {self.size})"
            )
        return self.sandboxes[index]

    async def cleanup(self) -> None:
        """Stop and delete all sandboxes. Ignores errors."""
        await asyncio.gather(*(s.cleanup() for s in self.sandboxes))


async def create_fleet(
    http: SmolvmHttpClient,
    prefix: str,
    count: int,
    opts: CreateSandboxOptions | None = None,
) -> SandboxFleet:
    """Create a fleet of sandboxes. Names: {prefix}-0, {prefix}-1, etc."""
    if opts is None:
        opts = CreateSandboxOptions(cpus=1, memory_mb=512, network=True)

    sandboxes: list[Sandbox] = []

    for i in range(count):
        name = f"{prefix}-{i}"

        # Cleanup any leftover from previous runs
        try:
            await http.stop_sandbox(name)
        except Exception:
            pass
        try:
            await http.delete_sandbox(name)
        except Exception:
            pass

        # Create
        req: dict[str, Any] = {
            "name": name,
            "resources": {
                "cpus": opts.cpus,
                "memory_mb": opts.memory_mb,
                "network": opts.network,
            },
        }
        if opts.overlay_gb is not None:
            req["resources"]["overlay_gb"] = opts.overlay_gb
        if opts.storage_gb is not None:
            req["resources"]["storage_gb"] = opts.storage_gb
        if opts.allowed_domains is not None:
            req["resources"]["allowed_domains"] = opts.allowed_domains
        if opts.mounts:
            req["mounts"] = [
                {"source": m.source, "target": m.target, "readonly": m.readonly}
                for m in opts.mounts
            ]
        if opts.ports:
            req["ports"] = [
                {"host": p.host, "guest": p.guest} for p in opts.ports
            ]
        if opts.init_commands is not None:
            req["init_commands"] = opts.init_commands
        if opts.default_user is not None:
            req["default_user"] = opts.default_user
        if opts.from_starter is not None:
            req["from_starter"] = opts.from_starter

        await http.create_sandbox(req)
        sandboxes.append(Sandbox(name, http))

    # Start all in parallel
    await asyncio.gather(*(s.start() for s in sandboxes))

    return SandboxFleet(sandboxes, http)

File: sdk-ts/fleet.ts
/**
 * smolvm TypeScript SDK — Fleet
 *
 * Multi-machine orchestration. Creates N machines, executes commands
 * across all of them in parallel, and cleans up.
 */

import { SmolvmHttpClient } from "./client.ts";
import { Machine } from "./machine.ts";
import type { CreateMachineOptions, ExecOptions, ExecResult } from "./types.ts";

export class MachineFleet {
  public readonly machines: Machine[];

  constructor(
    machines: Machine[],
    private readonly http: SmolvmHttpClient,
  ) {
    this.machines = machines;
  }

  /** Number of machines in the fleet. */
  get size(): number {
    return this.machines.length;
  }

  /** All machine names. */
  get names(): string[] {
    return this.machines.map((s) => s.name);
  }

  /**
   * Execute a shell command across all machines in parallel.
   * Returns results in the same order as machines.
   */
  async execAll(cmd: string, opts?: ExecOptions): Promise<ExecResult[]> {
    return Promise.all(this.machines.map((s) => s.sh(cmd, opts)));
  }

  /**
   * Execute different commands on each machine in parallel.
   * Commands array must match fleet size.
   */
  async execEach(cmds: string[], opts?: ExecOptions): Promise<ExecResult[]> {
    if (cmds.length !== this.machines.length) {
      throw new Error(
        `Expected ${this.machines.length} commands, got ${cmds.length}`,
      );
    }
    return Promise.all(
      this.machines.map((s, i) => s.sh(cmds[i], opts)),
    );
  }

  /**
   * Execute a raw command array across all machines in parallel.
   */
  async execAllRaw(
    command: string[],
    opts?: ExecOptions,
  ): Promise<ExecResult[]> {
    return Promise.all(this.machines.map((s) => s.exec(command, opts)));
  }

  /** Get a specific machine by index. */
  at(index: number): Machine {
    const s = this.machines[index];
    if (!s) throw new Error(`Fleet index ${index} out of range (size: ${this.size})`);
    return s;
  }

  /** Stop and delete all machines. Ignores errors. */
  async cleanup(): Promise<void> {
    await Promise.all(this.machines.map((s) => s.cleanup()));
  }
}

/**
 * Create a fleet of machines.
 * Names are auto-generated as `{prefix}-0`, `{prefix}-1`, etc.
 */
export async function createFleet(
  http: SmolvmHttpClient,
  prefix: string,
  count: number,
  opts?: CreateMachineOptions,
): Promise<MachineFleet> {
  const machines: Machine[] = [];

  for (let i = 0; i < count; i++) {
    const name = `${prefix}-${i}`;

    // Cleanup any leftover from previous runs
    try {
      await http.stopMachine(name);
    } catch { /* ignore */ }
    try {
      await http.deleteMachine(name);
    } catch { /* ignore */ }

    // Create
    await http.createMachine({
      name,
      mounts: opts?.mounts,
      ports: opts?.ports,
      resources: {
        cpus: opts?.cpus ?? 1,
        memoryMb: opts?.memoryMb ?? 512,
        network: opts?.network ?? true,
        overlayGb: opts?.overlay_gb,
        storageGb: opts?.storage_gb,
        allowedDomains: opts?.allowed_domains,
        allowedCidrs: opts?.allowed_cidrs,
      },
      init_commands: opts?.init_commands,
      default_user: opts?.default_user,
      from_starter: opts?.fromStarter,
    });

    machines.push(new Machine(name, http));
  }

  // Start all in parallel
  await Promise.all(machines.map((s) => s.start()));

  return new MachineFleet(machines, http);
}
